block/manager.go

```diff
@@ -0,0 +1,283 @@
+			newHeight := header.Height
+			currentHeight := m.store.Height()
+			if newHeight > currentHeight {
+				atomic.StoreUint64(&m.syncTarget, newHeight)
+				m.retrieveCh <- newHeight
+			}
+		case block := <-m.blockInCh:
+			m.logger.Debug("block body retrieved from DALC",
+				"height", block.Header.Height,
+				"hash", block.Hash(),
+			)
+			m.syncCache[block.Header.Height] = block
+			currentHeight := m.store.Height() // TODO(tzdybal): maybe store a copy in memory
+			b1, ok1 := m.syncCache[currentHeight+1]
+			b2, ok2 := m.syncCache[currentHeight+2]
+			if ok1 && ok2 {
+				newState, _, err := m.executor.ApplyBlock(ctx, m.lastState, b1)
+				if err != nil {
+					m.logger.Error("failed to ApplyBlock", "error", err)
+					continue
+				}
+				err = m.store.SaveBlock(b1, &b2.LastCommit)
+				if err != nil {
+					m.logger.Error("failed to save block", "error", err)
+					continue
+				}
+
+				m.lastState = newState
+				err = m.store.UpdateState(m.lastState)
+				if err != nil {
+					m.logger.Error("failed to save updated state", "error", err)
+					continue
+				}
+				delete(m.syncCache, currentHeight+1)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (m *Manager) RetrieveLoop(ctx context.Context) {
+	for {
+		select {
+		case <-m.retrieveCh:
+			target := atomic.LoadUint64(&m.syncTarget)
+			for h := m.store.Height() + 1; h <= target; h++ {
+				m.logger.Debug("trying to retrieve block from DALC", "height", h)
+				m.mustRetrieveBlock(ctx, h)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (m *Manager) mustRetrieveBlock(ctx context.Context, height uint64) {
+	// TOOD(tzdybal): extract configuration option
+	maxRetries := 10
+
+	for r := 0; r < maxRetries; r++ {
+		err := m.fetchBlock(ctx, height)
+		if err == nil {
+			return
+		}
+		// TODO(tzdybal): configuration option
+		// TODO(tzdybal): exponential backoff
+		time.Sleep(100 * time.Millisecond)
+	}
+	// TODO(tzdybal): this is only temporary solution, for MVP
+	panic("failed to retrieve block with DALC")
+}
+
+func (m *Manager) fetchBlock(ctx context.Context, height uint64) error {
+	var err error
+	blockRes := m.retriever.RetrieveBlock(height)
+	switch blockRes.Code {
+	case da.StatusSuccess:
+		m.blockInCh <- blockRes.Block
+	case da.StatusError:
+		err = fmt.Errorf("failed to retrieve block: %s", blockRes.Message)
+	case da.StatusTimeout:
+		err = fmt.Errorf("timeout during retrieve block: %s", blockRes.Message)
+	}
+	return err
+}
+
+func (m *Manager) getRemainingSleep(start time.Time) time.Duration {
+	publishingDuration := time.Since(start)
+	sleepDuration := m.conf.BlockTime - publishingDuration
+	if sleepDuration < 0 {
+		sleepDuration = 0
+	}
+	return sleepDuration
+}
+
+func (m *Manager) publishBlock(ctx context.Context) error {
+	var lastCommit *types.Commit
+	var err error
+	newHeight := m.store.Height() + 1
+
+	// this is a special case, when first block is produced - there is no previous commit
+	if newHeight == uint64(m.genesis.InitialHeight) {
```

Comment:
Shouldn't this be the current height, not the new height?

Reply:
It's the height calculated for newly created block / block to be created, thus `newHeight`.